File: apps/android/app/src/main/java/com/xueqing/app/infrastructure/remote/LearningReadProviderSupport.kt
package com.xueqing.app.infrastructure.remote

import com.xueqing.app.application.learning.LearningReadFailure
import com.xueqing.app.application.learning.LearningReadFailureKind
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

internal object LearningReadProviderSupport {
    fun mapFailure(
        statusCode: Int,
        body: String,
        json: Json,
    ): LearningReadFailure {
        if (statusCode == 401) {
            return LearningReadFailure(
                LearningReadFailureKind.AuthenticationRequired,
                "XQ_AUTH_REQUIRED",
            )
        }

        val message = providerMessage(body, json)
        if (message == "XQ_AUTH_REQUIRED" || message == "XQ_ACTOR_NOT_FOUND") {
            return LearningReadFailure(
                LearningReadFailureKind.AuthenticationRequired,
                message,
            )
        }

        if (
            message == "XQ_ACTOR_DISABLED" ||
            message == "XQ_TEACHING_CONTEXT_REQUIRED" ||
            message == "XQ_TEACHING_CONTEXT_UNAVAILABLE"
        ) {
            return LearningReadFailure(LearningReadFailureKind.AccessDenied, message)
        }

        if (
            message == "XQ_CASE_PRIMARY_ACTION_INVARIANT" ||
            message == "XQ_INVALID_ORGANIZATION_TIME_ZONE" ||
            message == "XQ_BUSINESS_DATE_INPUT_REQUIRED"
        ) {
            return LearningReadFailure(LearningReadFailureKind.ServerInvariant, message)
        }

        if (statusCode == 429 || statusCode >= 500) {
            return LearningReadFailure(
                LearningReadFailureKind.Transient,
                message ?: "HTTP_$statusCode",
            )
        }

        return LearningReadFailure(
            LearningReadFailureKind.InvalidResponse,
            message ?: "HTTP_$statusCode",
        )
    }

    private fun providerMessage(
        body: String,
        json: Json,
    ): String? = runCatching {
        val root = json.parseToJsonElement(body) as? JsonObject
            ?: error("Provider error body must be an object.")
        val message = root["message"] as? JsonPrimitive
            ?: error("Provider error body has no message.")
        require(message.isString)
        message.content
    }.getOrNull()
}

File: apps/windows/src/Xueqing.Windows.Infrastructure/Remote/LearningReadProviderSupport.cs
using System.Net;
using System.Text.Json;
using Xueqing.Windows.Core.Models;

namespace Xueqing.Windows.Infrastructure.Remote;

internal static class LearningReadProviderSupport
{
    public static Uri BuildRpcUri(Uri projectUri, string rpcPath)
    {
        if (!projectUri.IsAbsoluteUri)
        {
            throw new ArgumentException("Reference-provider URI must be absolute.", nameof(projectUri));
        }

        if (projectUri.Scheme != Uri.UriSchemeHttps &&
            !(projectUri.Scheme == Uri.UriSchemeHttp && projectUri.IsLoopback))
        {
            throw new ArgumentException(
                "Reference-provider URI must use HTTPS; loopback HTTP is allowed for development tests.",
                nameof(projectUri));
        }

        var normalized = projectUri.AbsoluteUri.EndsWith("/", StringComparison.Ordinal)
            ? projectUri
            : new Uri(projectUri.AbsoluteUri + "/", UriKind.Absolute);
        return new Uri(normalized, rpcPath);
    }

    public static (LearningReadFailureKind Kind, string Code) MapFailure(
        HttpStatusCode statusCode,
        string body)
    {
        if (statusCode == HttpStatusCode.Unauthorized)
        {
            return (LearningReadFailureKind.AuthenticationRequired, "XQ_AUTH_REQUIRED");
        }

        var message = TryReadProviderMessage(body);
        if (message is "XQ_AUTH_REQUIRED" or "XQ_ACTOR_NOT_FOUND")
        {
            return (LearningReadFailureKind.AuthenticationRequired, message);
        }

        if (message is "XQ_ACTOR_DISABLED" or
            "XQ_TEACHING_CONTEXT_REQUIRED" or
            "XQ_TEACHING_CONTEXT_UNAVAILABLE")
        {
            return (LearningReadFailureKind.AccessDenied, message);
        }

        if (message is "XQ_CASE_PRIMARY_ACTION_INVARIANT" or
            "XQ_INVALID_ORGANIZATION_TIME_ZONE" or
            "XQ_BUSINESS_DATE_INPUT_REQUIRED")
        {
            return (LearningReadFailureKind.ServerInvariant, message);
        }

        if ((int)statusCode == 429 || (int)statusCode >= 500)
        {
            return (LearningReadFailureKind.Transient, message ?? $"HTTP_{(int)statusCode}");
        }

        return (LearningReadFailureKind.InvalidResponse, message ?? $"HTTP_{(int)statusCode}");
    }

    private static string? TryReadProviderMessage(string body)
    {
        try
        {
            using var document = JsonDocument.Parse(body);
            var root = document.RootElement;
            if (root.ValueKind == JsonValueKind.Object &&
                root.TryGetProperty("message", out var message) &&
                message.ValueKind == JsonValueKind.String)
            {
                return message.GetString();
            }
        }
        catch (JsonException)
        {
        }

        return null;
    }
}
